### ISSUE TITLE
values not resolving when including the same sub-sub-chart multiple times

### ISSUE THREAD

**Issue by jeroenjacobs79** [helm/helm#9789]
Output of `helm version`:

`version.BuildInfo{Version:"v3.6.0", GitCommit:"7f2df6467771a75f5646b7f12afb408590ed1755", GitTreeState:"dirty", GoVersion:"go1.16.4"}`

Output of `kubectl version`:

```
kubectl version
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.1", GitCommit:"5e58841cce77d4bc13713ad2b91fa0d961e69192", GitTreeState:"clean", BuildDate:"2021-05-12T14:11:29Z", GoVersion:"go1.16.3", Compiler:"gc", Platform:"darwin/amd64"}
```

Cloud Provider/Platform (AKS, GKE, Minikube etc.): N/A

# Situation

We created a base chart for our developers, to make it easier for them to create their own application charts. They just include the base chart as a dependency, and customise some variables. Later on, we create an "umbrella" chart, which includes the application chart as a dependency.

# Base helm chart

I'm going to reduce the base chart to something simple, to make it easier to show the problem.

Chart.yaml:
```
apiVersion: v2
name: base
description: Base helm chart for all applications
version: 0.1.0
```

values.yaml:
```
# -- name of your application
appName: "my-app"
serviceAccount:
  create: true
```

templates/_helpers.tpl:
```

{{- define "application.fullname" -}}
{{- $name := .Values.appName }}
{{- if contains $name .Release.Name }}
    {{- .Release.Name | trunc 63 | trimSuffix "-" }}
{{- else }}
{{- printf "%s-%s" .Release.Name $name | trunc 63 | trimSuffix "-" }}
{{- end }}
{{- end }}

{{- define "application.serviceAccountName" -}}
{{- if .Values.serviceAccount.create }}
{{- default (include "application.fullname" .) .Values.serviceAccount.name }}
{{- else }}
{{- default "default" .Values.serviceAccount.name }}
{{- end }}
{{- end }}
```

templates/serviceaccount.yaml
```
{{- if .Values.serviceAccount.create -}}
apiVersion: v1
kind: ServiceAccount
metadata:
  name: {{ include "application.serviceAccountName" . }}
  labels:
    {{- include "application.labels" . | nindent 4 }}
    {{- with .Values.labels }}
    {{- toYaml . | nindent 4 }}
    {{- end }}
    {{- with .Values.serviceAccount.labels }}
    {{- toYaml . | nindent 4 }}
    {{- end }}
  {{- if or .Values.annotations  .Values.serviceAccount.annotations }}
  annotations:
    {{- with .Values.annotations }}
    {{- toYaml . | nindent 4 }}
    {{- end }}
    {{- with .Values.serviceAccount.annotations }}
    {{- toYaml . | nindent 4 }}
    {{- end }}
  {{- end }}
{{- end }}
```

Nothing special going on here, we just create a SA here. The default name of the app is "my-app", we will encounter this one again...

# Application chart

Again, I give a shorter example just to illustrate the issue

Chart.yaml:
```
apiVersion: v2
name: super-application
description: app helm chart for super application
version: 0.1.0
dependencies:
  - name: base
    repository: <path-to-repo>
    version: 0.1.0
    alias: app
```

values.yaml:
```
app:
  # -- name of your application
  appName: "super-application"
```

So far, no issue. The final render of the service account looks more or less like this:

```
apiVersion: v1
kind: ServiceAccount
metadata:
  name: RELEASE-NAME-super-application
  labels:
    app.kubernetes.io/managed-by: Helm
```

# Umbrella chart

And now, for the umbrella chart, and to illustrate the issue...

Chart.yaml
```
apiVersion: v2
name: umbrella
description: helm chart for launching all applications
version: 0.1.0
dependencies:
  - name: super-application
    repository: <path-to-repo>
    version: 0.1.0
    alias: app-1
    condition: app-1.enabled
  - name: super-application
    repository: <path-to-repo>
    version: 0.1.0
    alias: app-2
    condition: app-2.enabled
```

values.yaml:

```
app-1:
  enabled: true
  app:
    appName: "app-1"

app-1:
  enabled: true
  app:
    appName: "app-2"
```

To be clear: As long we enable just one application, there is no issue! The issue only occurs when BOTH applications are enabled in the values file.

This is what it SHOULD be (according to my understanding):

```
apiVersion: v1
kind: ServiceAccount
metadata:
  name: RELEASE-NAME-app-1
  labels:
    app.kubernetes.io/managed-by: Helm
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: RELEASE-NAME-app-2
  labels:
    app.kubernetes.io/managed-by: Helm
```

However, this what it ACTUALLY templates to:

```
apiVersion: v1
kind: ServiceAccount
metadata:
  name: RELEASE-NAME-app-1
  labels:
    app.kubernetes.io/managed-by: Helm
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: RELEASE-NAME-my-app
  labels:
    app.kubernetes.io/managed-by: Helm
```

One of the appNames is getting evaluated to "my-app", instead of "app-2", which is not the expected result.

Is this a bug, due to the fact that the same subchart is included by all applications?

**Comment by mattfarina** [comment]
Thank you for providing such a detailed explanation with examples. This makes it much easier to figure out what it doing on. I will try to provide a sufficient response.

When Helm installs a chart it installs all of the dependent charts together. When it renders the templates it does so in one shared sandbox. This is why template functions need to be namespaced with the chart name as names can have collisions and only one is used. This is the problem you are encountering.

This was a design decision a long time ago with Helm. In Helm v3 this is a works as expected situation and a limitation.

Helm is a package manager like apt, yum, homebrew, etc. What we usually suggest is that people in your situation couple Helm with a deployment tool. The equivalent of a Chef or Ansible in the way they use apt or yup packages. Those tools end up separating chart installs in a manner where they are installed separately which means a separate template rendering sandbox. For this there are tools like fleet, the Flux Helm operator, and Helmfile.

I have personally been working on a way to handle some of this differently for Helm with [Hypper](https://hypper.io/).

Does this help?

**Comment by jeroenjacobs79** [comment]
Hi,

I'm not sure what you mean with "template functions need to be namespaced with the chart name", do you have an example of this?

Can this be circumvented by not using template helpers?

Also, if there is collision, I would expect that they would both render to the same result (either "app-1" or "app-2"), not the default of the sub-subchart. I do not understand why this only affects sub-subcharts, and not sub-charts either, eg: when I use two dependencies on the base-chart instead of the app-chart with a different alias, the issue doesn't occur either. It only seems to happen with sub-subcharts.

Switching tools is not an options I'm afraid. Our intention is to use ArgoCD, which basically just uses `helm template` under the hood.

**Comment by mattfarina** [comment]
> I'm not sure what you mean with "template functions need to be namespaced with the chart name", do you have an example of this?

In a chart you'll see functions defined using something like `{{- define "application.fullname" -}}`. Here the `application` provides a namespace. This ways `application.fullname` doesn't conflict with `application2.fullname`. If they both used `fullname` it would be a conflict in the same sandbox the templates are rendered in.

> Can this be circumvented by not using template helpers?

Unfortunately, it can't be. Everything is in a shared sandbox for rendering.

> Also, if there is collision, I would expect that they would both render to the same result (either "app-1" or "app-2"), not the default of the sub-subchart. I do not understand why this only affects sub-subcharts, and not sub-charts either, eg: when I use two dependencies on the base-chart instead of the app-chart with a different alias, the issue doesn't occur either. It only seems to happen with sub-subcharts.

It should affect everything being rendered. I would need to see the whole real chart to pick it apart and find out what's going on.

> Switching tools is not an options I'm afraid. Our intention is to use ArgoCD, which basically just uses helm template under the hood.

I understand.

While I don't know how setup and you may have already looked at this, you might consider breaking up your macro app into smaller units for deployment via Argo.

**Comment by manzsolutions-lpreis** [comment]
I agree with @jeroenjacobs79 here that the behaviour is not what one would expect from a sandboxing issue. I think we are observing a variation of this issue.

We have a generic chart for Springboot and we added a custom Redis chart as an *optional* sub-chart to it. Another app with its own chart also uses the Redis chart plus the Springboot chart for two applications but with disabled Redis.

```
apiVersion: v2
name: springboot
description: generic springboot helm chart
version: 0.1.0
dependencies:
  - name: our-redis
    repository: <path-to-repo>
    condition: redis.enabled
    version: 0.1.0
    alias: redis
```

```
apiVersion: v2
name: main-app
description: app helm chart for main application
version: 0.1.0
dependencies:
  - name: springboot
    repository: <path-to-repo>
    version: 0.1.0
    alias: sub-app-1
  - name: springboot
    repository: <path-to-repo>
    version: 0.1.0
    alias: sub-app-2
  - name: our-redis
    repository: <path-to-repo>
    version: 0.1.0
    alias: redis
```

Both Springboot subcharts render correctly as does the redis chart. But the *second* Springboot-chart also renders its Redis-dependency, with varying success. The config which directly references `.Values.redis` is almost identical to the one rendered by main-app.redis, but the pull secret which uses `{{ with .Values.redis.credentials }}` remains empty and the alias is ignored. Which is how we ran into this bug since it produces invalid manifests.

It gets even better: If one adds faulty values for the *disabled* sub-sub-chart main-app.app-1.redis suddenly the main-app.redis template rendering complains. Also, and that’s my favourite part, if you remove any one of the two Springboot chart dependencies keeping the other the mis-rendered templates are gone, too.

Long story short: This cannot just be a namespacing issue. After all we do not even want to render the sub-sub-chart but the condition doesn't take and also ignores the alias.

**Comment by jeroenjacobs79** [comment]
This issue still exists. I still believe this needs to be fixed, as the current behaviour is counter-intuitive and confusing.

**Comment by jeroenjacobs79** [comment]
I assume not attempts will be made to fix this odd behaviour?

I can live with that, but it would be nice to get some feedback from the developers on this issue.

**Comment by renat1sakenov** [comment]
We're also affected by this problem. As a workaround one could copy the inital Chart and add "-2" to its name, but hopefully that would only be temporary.
